share/scripts/sdcdb.tcl

```diff
@@ -0,0 +1,795 @@
+        variable $arrayname
+        fix_last_address $arrayname
+        fix_blank_spaces $arrayname
+    }
+    variable a_files
+    foreach filename [array names a_files] {
+        set arrayname a_[file rootname $filename]2addr
+        variable $arrayname
+        fix_last_address $arrayname
+        fix_blank_spaces $arrayname
+    }
+    fix_blank_spaces addr2file
+    return
+}
+
+proc normal_glob {dir pattern} {
+    glob -nocomplain -type f -directory $dir $pattern
+}
+
+proc recursive_glob {dir pattern} {
+    set result [list]
+    foreach file [glob -nocomplain -directory $dir $pattern] {
+        lappend result $file
+    }
+    foreach subdir [glob -nocomplain -directory $dir *] {
+        if {[file isdirectory $subdir]} {
+            lappend result {*}[recursive_glob $subdir $pattern]
+        }
+    }
+    return $result
+}
+
+proc add_files_to_database {arrayname files} {
+    variable $arrayname
+    foreach path $files {
+        set filename [file tail $path]
+        if {[array get $arrayname $filename] ne {}} {
+            debug_out "file '$filename' already registered in $arrayname, new entry ignored."
+        }
+        set ${arrayname}($filename) $path
+    }
+}
+
+proc sdcdb_list {args} {
+    # parameter pattern 1: file:beginLine-endLine
+    set pattern1 {([^:]+):(\d+)(-(\d+))?}
+    # parameter pattern 2: file:functionName
+    set pattern2 {([^:]+):(\S+)}
+    # parameter pattern 3: functionName
+    set pattern3 {(\S+)}
+    set arg [lindex $args 0]
+    set match [regexp -inline $pattern1 $arg]
+    if {[llength $match] == 5} {
+        lassign $match {} file start {} end
+        return [list_file $file $start $end 0 1]
+    }
+    set match [regexp -inline $pattern2 $arg]
+    if {[llength $match] == 3} {
+        lassign $match {} file funcname
+        return [list_fun $file $funcname]
+    }
+    set match [regexp -inline $pattern3 $arg]
+    if {[llength $match] == 2} {
+        lassign $match {} funcname
+        return [list_fun {} $funcname]
+    }
+    list_pc
+}
+
+proc list_file {file begin {end {}} {focus 0} {showerror 0}} {
+    if {[file extension $file] eq ".c"} {
+        set files c_files
+    } elseif {[file extension $file] eq [ASM]} {
+        set files a_files
+    } else {
+        error "Unknown or unspecified file extension"
+    }
+    variable $files
+    set record [array get $files $file]
+    if {$record eq {}} {
+        if {$showerror} {
+            error "file '$file' not found in database, add a directory that contains such file with 'sdcdb add <dir>'"
+        }
+        puts "$file: not found"
+        return
+    }
+    debug_out "opening file [lindex $record 1]..."
+    set fh [open [lindex $record 1] r]
+    set pos 0
+    # 10 lines by default
+    if {$end eq {}} { set end [expr {$begin + 9}] }
+    while {[gets $fh line] >= 0} {
+        incr pos
+        if {$pos >= $begin} {
+            puts "[format %-5d $pos][expr {$pos == $focus ? "*" : ":"}]    $line"
```

Comment:
Same here, should these be accumulated in a multi-line string, and then returned?